File: KoreaCredit/app/src/main/java/lib/datamanager/DataManager.java
package lib.datamanager;





import java.io.BufferedOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import java.net.URL;

import android.os.AsyncTask;
import android.util.Log;


import org.json.JSONException;
import org.json.JSONObject;

import java.io.ByteArrayOutputStream;
import java.io.DataOutputStream;
import java.io.InputStream;
import java.net.HttpURLConnection;






public class DataManager implements DataLoaderDelegate
{
	

	
	private static final String TAG = "dataManager";

    private static final String RETURN_TYPE_JSON="JSON";
    private static final String RETURN_TYPE_TXT="TXT";
	
	

    public JsonManagerDelegate jsonDelegate;
    public DataManagerDelegate dataDelegate;
    private Map<String,DataLoader> loaderA;

 	public String type,returnType;

    
    public DataManager (String _type)
	{

		 type=_type;

         loaderA =new HashMap<String,DataLoader>();
		
		 
	}
    public void setOnJsonDelegate (JsonManagerDelegate _jsonDelegate)
    {
        jsonDelegate=_jsonDelegate;
        returnType=RETURN_TYPE_JSON;

    }
    public void setOnDataDelegate (DataManagerDelegate _dataDelegate)
    {
        dataDelegate=_dataDelegate;
        returnType=RETURN_TYPE_TXT;

    }
    public void removeAllLoader() { 
		
    	loaderA.clear();
	} 
    
    public void removeLoader (String dataUrl) { 
    	DataLoader loader=loaderA.get(dataUrl);
		if(loader!=null){
			loader.cancel(false);
			loaderA.remove(dataUrl);
		}
	} 
	public void loadData(String dataUrl) { 
		
		removeLoader (dataUrl);
		DataLoader loader=new DataLoader();
		loaderA.put(dataUrl, loader);
		loader.start(this,null,type,false,null,"",null,"application/x-www-form-urlencoded");
		loader.execute(dataUrl);
		
	} 
    public void loadData(String dataUrl,Map<String,String> param) {
		
    	
    	removeLoader (dataUrl);
    	DataLoader loader=new DataLoader();
		loaderA.put(dataUrl, loader);
		loader.start(this,param,type,false,null,"",null,"application/x-www-form-urlencoded");
		loader.execute(dataUrl);
		
	}
    public void loadData(String dataUrl,Map<String,String> param,ArrayList<FileObject> files) {


        removeLoader (dataUrl);
        DataLoader loader=new DataLoader();
        loaderA.put(dataUrl, loader);
        loader.start(this,param,"POST",true,files,"",null,"multipart/form-data");
        loader.execute(dataUrl);

    }
    public void loadData(String dataUrl,Map<String,String> param,ArrayList<FileObject> files,String boundary) {


        removeLoader (dataUrl);
        DataLoader loader=new DataLoader();
        loaderA.put(dataUrl, loader);
        loader.start(this, param, "POST", true, files, boundary,null,"multipart/form-data");
        loader.execute(dataUrl);

    }
    public void loadData(String dataUrl,Map<String,String> param,ArrayList<FileObject> files,String boundary,Map<String,String> headers) {


        removeLoader (dataUrl);
        DataLoader loader=new DataLoader();
        loaderA.put(dataUrl, loader);
        loader.start(this, param, "POST", true, files, boundary, headers,"multipart/form-data");
        loader.execute(dataUrl);

    }
    public void loadData(String dataUrl,Map<String,String> param,ArrayList<FileObject> files,String boundary,Map<String,String> headers,String contentType) {


        removeLoader (dataUrl);
        DataLoader loader=new DataLoader();
        loaderA.put(dataUrl, loader);
        loader.start(this, param, "POST", true, files, boundary, headers,contentType);
        loader.execute(dataUrl);

    }
    public void destory() {
		
		
		removeAllLoader();
        dataDelegate=null;
        jsonDelegate=null;
		loaderA=null;
		
	} 
	
	public void onCompleted(String dataUrl,String result) {
        removeLoader (dataUrl);
        if(dataDelegate!=null){
            dataDelegate.onDataCompleted(this,dataUrl, result);

		};
        if(jsonDelegate!=null){

            try {

                Log.i(TAG, "result : "+result);
                JSONObject responseJSON = new JSONObject(result);
                jsonDelegate.onJsonCompleted(this, dataUrl, responseJSON);
            } catch (JSONException e) {
                Log.e(TAG, "Error in responseJSON parser ", e);
                jsonDelegate.onJsonParseErr(this, dataUrl);

            }


        };

	} 
	public void onLoadErr(String dataUrl) {
        removeLoader (dataUrl);
		if(dataDelegate!=null){
            dataDelegate.onDataLoadErr(this,dataUrl);

		};
        if(jsonDelegate!=null){
            jsonDelegate.onJsonLoadErr(this, dataUrl);

        };

    } 
	
	
	public  interface DataManagerDelegate {
	    void onDataCompleted(DataManager manager,String path,String result);
	    void onDataLoadErr(DataManager manager,String path);

	}
    public  interface JsonManagerDelegate {
        void onJsonCompleted(DataManager manager,String path,JSONObject result);
        void onJsonParseErr(DataManager manager,String path);
        void onJsonLoadErr(DataManager manager,String path);

    }
}


class DataLoader extends AsyncTask<String, Integer, String> 
{
	
	  private static final String TAG = "dataLoader";
	  private String path,boundary = "^******^";;
	  private DataLoaderDelegate delegate;
	  private Map<String,String> urlParam,headers;
      private ArrayList<FileObject> files;
	  private String type,contentType;
      private boolean isMultipart;


	  public void start(DataLoaderDelegate _delegate,Map<String,String> param,String _type,
                        boolean _isMultipart,ArrayList<FileObject> _files,String _boundary,
                        Map<String,String> _headers,String _contentType) {
		    path = "";
	        delegate = _delegate;
	        urlParam=param;
            contentType=_contentType;
            headers=_headers;
            if(_boundary.equals("")==false) {
                boundary=_boundary;
            }
            files=_files;
            isMultipart=_isMultipart;
	        type=_type;
	        Log.i(TAG, "DataLoader START "+type+" mp:"+isMultipart);
	  }



	  protected String doInBackground(String... params) {
		    Log.i(TAG, "BEGIN ParserThread");
		    
		    path = params[0];
            HttpURLConnection uc=null;
            InputStream is=null;
		    String responseStr="";
		    
            try {

                    URL text = new URL(path);
                    uc = (HttpURLConnection) text.openConnection();

                    uc.setDoInput(true);
                    if (type.equals("POST")||isMultipart==true) {
                        uc.setDoOutput(true);
                    }
                    uc.setUseCaches(false);
                    uc.setRequestMethod(type);
                    uc.setConnectTimeout(10000);  //
                    uc.setAllowUserInteraction(true);

                    if(headers!=null){
                        Object[] keySet = headers.keySet().toArray();
                        String key;

                        for (int i = 0; i < keySet.length; ++i) {
                            key = (String) keySet[i];
                            uc.setRequestProperty(key, headers.get(key));
                            Log.i(TAG, "header = " + key + ":" + headers.get(key));
                        }

                    }


                    if(isMultipart==true){


                        String delimiter = "\r\n--" + boundary + "\r\n";
                        uc.setRequestProperty("Connection", "Keep-Alive");
                        uc.setRequestProperty("Content-Type", contentType+" ;boundary="
                                + boundary);


                        DataOutputStream wr = new DataOutputStream(new BufferedOutputStream(
                                uc.getOutputStream()));
                        if (urlParam != null) {


                            String write = DataLoaderUtil.getParamsString(urlParam, delimiter);
                            wr.writeUTF(write);

                            Log.i(TAG, "dataUrl=" + path + "?" + write);
                        } else {
                            Log.i(TAG, "dataUrl=" + path);

                        }

                        if(files!=null){
                            for(int i=0;i<files.size();++i){

                                Log.i(TAG, "file =" + files.get(i).key+" : "+files.get(i).name);
                                DataLoaderUtil.setFileBuffer(files.get(i), wr, delimiter);

                            }

                        }
                        wr.writeBytes(delimiter);
                        wr.flush();
                        wr.close();


                    }else{

                        uc.setRequestProperty("Content-type", contentType);

                        if (urlParam != null) {
                            StringBuffer sb = new StringBuffer();
                            Object[] keySet = urlParam.keySet().toArray();
                            String key;

                            for (int i = 0; i < keySet.length; ++i) {
                                key = (String) keySet[i];
                                if (i == (keySet.length - 1)) {
                                    sb.append(key).append("=").append(urlParam.get(key));
                                } else {
                                    sb.append(key).append("=").append(urlParam.get(key)).append("&");
                                }
                            }

                            DataOutputStream wr = new DataOutputStream(uc.getOutputStream());
                            String write = sb.toString();
                            wr.writeBytes(write);
                            wr.flush();
                            wr.close();
                            Log.i(TAG, "dataUrl=" + path + "?" + write);
                        } else {
                            Log.i(TAG, "dataUrl=" + path);

                        }
                    }





                    int responseCode = uc.getResponseCode();
                    if (responseCode == HttpURLConnection.HTTP_OK) {
                        is = uc.getInputStream();
                        //uc.disconnect();

                    }

            }catch(Exception e){

                 Log.e(TAG, "Error in parser run", e);
                 return null;

            }
            if(is!=null){

                ByteArrayOutputStream baos = new ByteArrayOutputStream();
                byte[] byteBuffer = new byte[1024];
                byte[] byteData = null;
                int nLength = 0;
                try {
                    while ((nLength = is.read(byteBuffer, 0, byteBuffer.length)) != -1) {
                        baos.write(byteBuffer, 0, nLength);
                    }
                } catch (IOException e) {
                    Log.e(TAG, "Error in InputStream read ", e);
                    return null;
                }
                byteData = baos.toByteArray();
                responseStr = new String(byteData);
                //Log.i(TAG, "response=" + responseStr);



            }
            if(uc!=null){
                uc.disconnect();
            }
            return responseStr;
	    }
	  
	  
			  
	    
	    protected void onPostExecute(String result) {
	       
	        if(result==null){
	        	if(delegate != null) delegate.onLoadErr(path);
	        }else{
	        	if(delegate != null) delegate.onCompleted(path, result);
	        }
	    	super.onPostExecute(result);
	    }
	    
	   

}//package

interface DataLoaderDelegate
{
    void onCompleted(String path,String result);
    void onLoadErr(String path);
 
}

File: Magellan/app/src/main/java/lib/loader/DataManager.kt
package lib.loader
import android.os.AsyncTask
import android.util.Log
import org.json.JSONException
import org.json.JSONObject
import java.io.*
import java.net.HttpURLConnection
import java.net.URL
import java.util.ArrayList
import java.util.HashMap


class DataManager(var type: String) : DataLoaderDelegate
{
    var jsonDelegate: JsonManagerDelegate? = null
    var dataDelegate: DataManagerDelegate? = null
    private var loaderA: MutableMap<String, DataLoader> = HashMap()
    var returnType: String = RETURN_TYPE_TXT

    init
    {

    }

    fun setOnJsonDelegate(_jsonDelegate: JsonManagerDelegate)
    {
        jsonDelegate = _jsonDelegate
        returnType = RETURN_TYPE_JSON
    }

    fun setOnDataDelegate(_dataDelegate: DataManagerDelegate)
    {
        dataDelegate = _dataDelegate
        returnType = RETURN_TYPE_TXT

    }

    fun removeAllLoader()
    {
        val iter = loaderA.keys.iterator()
        while (iter.hasNext()) loaderA[iter.next()]?.cancel(false)
        loaderA!!.clear()
    }

    fun removeLoader(dataUrl: String?) {
        val loader = loaderA[dataUrl]
        if (loader != null) {
            loader.cancel(false)
            loaderA.remove(dataUrl)
        }
    }

    fun loadData(dataUrl: String)
    {
        removeLoader(dataUrl)
        val loader = DataLoader()
        loaderA[dataUrl] = loader
        loader.start(this, null, type, null, "", null, "application/x-www-form-urlencoded")
        loader.execute(dataUrl)
    }

    fun loadData(dataUrl: String, param: Map<String, String>)
    {
        removeLoader(dataUrl)
        val loader = DataLoader()
        loaderA!![dataUrl] = loader
        loader.start(this, param, type, null, "", null, "application/x-www-form-urlencoded")
        loader.execute(dataUrl)
    }

    fun loadData(dataUrl: String, file: FileObject)
    {
        removeLoader(dataUrl)
        val loader = DataLoader()
        loaderA!![dataUrl] = loader
        val files = ArrayList<FileObject>()
        files.add(file)
        loader.start(this, null, "POST", files, "", null, "form-data")
        loader.execute(dataUrl)
    }

    fun loadData(dataUrl: String, param: Map<String, String>, files: ArrayList<FileObject>)
    {
        removeLoader(dataUrl)
        val loader = DataLoader()
        loaderA!![dataUrl] = loader
        loader.start(this, param, "POST", files, "", null, "multipart/form-data")
        loader.execute(dataUrl)
    }

    fun loadData(dataUrl: String, param: Map<String, String>, files: ArrayList<FileObject>, boundary: String)
    {
        removeLoader(dataUrl)
        val loader = DataLoader()
        loaderA!![dataUrl] = loader
        loader.start(this, param, "POST", files, boundary, null, "multipart/form-data")
        loader.execute(dataUrl)
    }

    fun loadData(dataUrl: String, param: Map<String, String>, files: ArrayList<FileObject>, boundary: String, headers: Map<String, String>)
    {
        removeLoader(dataUrl)
        val loader = DataLoader()
        loaderA!![dataUrl] = loader
        loader.start(this, param, "POST", files, boundary, headers, "multipart/form-data")
        loader.execute(dataUrl)
    }

    fun loadData(dataUrl: String, param: Map<String, String>, files: ArrayList<FileObject>, boundary: String, headers: Map<String, String>, contentType: String)
    {
        removeLoader(dataUrl)
        val loader = DataLoader()
        loaderA!![dataUrl] = loader
        loader.start(this, param, "POST", files, boundary, headers, contentType)
        loader.execute(dataUrl)
    }

    fun destory()
    {
        removeAllLoader()
        dataDelegate = null
        jsonDelegate = null
    }

    override fun onCompleted(dataUrl: String?, result: String)
    {
        removeLoader(dataUrl)
        dataDelegate?.onDataCompleted(this, dataUrl, result)
        if (jsonDelegate == null) return
        try
        {
            Log.i(TAG, "result : $result")
            val responseJSON = JSONObject(result)
            jsonDelegate!!.onJsonCompleted(this, dataUrl, responseJSON)
        }
        catch (e: JSONException)
        {
            Log.e(TAG, "Error in responseJSON parser ", e)
            jsonDelegate!!.onJsonParseErr(this, dataUrl)

        }
    }

    override fun onLoadErr(dataUrl: String?)
    {
        removeLoader(dataUrl)
        dataDelegate?.onDataLoadErr(this, dataUrl)
        jsonDelegate?.onJsonLoadErr(this, dataUrl)
    }


    interface DataManagerDelegate
    {
        fun onDataCompleted(manager: DataManager, path: String?, result: String){}
        fun onDataLoadErr(manager: DataManager, path: String?){}
    }

    interface JsonManagerDelegate
    {
        fun onJsonCompleted(manager: DataManager, path: String?, result: JSONObject){}
        fun onJsonParseErr(manager: DataManager, path: String?){}
        fun onJsonLoadErr(manager: DataManager, path: String?){}
    }

    companion object
    {
        private val TAG = "dataManager"
        private val RETURN_TYPE_JSON = "JSON"
        private val RETURN_TYPE_TXT = "TXT"
    }
}


internal class DataLoader : AsyncTask<String, Int, String>()
{
    private var path: String? = null
    private var boundary = "gc0p4Jq0M2Yt08jU534c0p"
    private var delegate: DataLoaderDelegate? = null
    private var urlParam: Map<String, String>? = null
    private var headers: Map<String, String>? = null
    private var files: ArrayList<FileObject>? = null
    private var type: String? = null
    private var contentType: String? = null
    private var isMultipart = false

    fun start(_delegate: DataLoaderDelegate, param: Map<String, String>?, _type: String,
              _files: ArrayList<FileObject>?, _boundary: String,
              _headers: Map<String, String>?, _contentType: String)
    {
        path = ""
        delegate = _delegate
        urlParam = param
        contentType = _contentType
        if (contentType === "multipart/form-data") {
            isMultipart = true
        }

        headers = _headers
        if (_boundary == "" == false) boundary = _boundary
        files = _files
        type = _type
        Log.i(TAG, "DataLoader START $type mp:$isMultipart")
    }


    override fun doInBackground(vararg params: String): String?
    {
        Log.i(TAG, "BEGIN ParserThread")
        path = params[0]
        var uc: HttpURLConnection? = null
        var ism: InputStream? = null
        var responseStr = ""

        try {

            val text = URL(path!!)
            uc = text.openConnection() as HttpURLConnection

            uc.doInput = true
            if (type == "POST" || isMultipart == true) {
                uc.doOutput = true
            }
            uc.useCaches = false
            uc.requestMethod = type!!
            uc.connectTimeout = 10000  //
            uc.allowUserInteraction = true

            if (headers != null) {
                val keySet = headers!!.keys.toTypedArray()
                var key: String

                for (i in keySet.indices) {
                    key = keySet[i]
                    uc.setRequestProperty(key, headers!![key])
                    Log.i(TAG, "header = " + key + ":" + headers!![key])
                }

            }


            if (isMultipart == true) {

                val delimiter = "\r\n--$boundary\r\n"
                uc.setRequestProperty("Connection", "Keep-Alive")
                val ctype = "$contentType ;boundary=$boundary"
                uc.setRequestProperty("Content-Type", ctype)
                //uc.setRequestProperty("enctype", contentType);
                Log.i(TAG, "Content-Type =$ctype")
                val wr = DataOutputStream(BufferedOutputStream(
                        uc.outputStream))

                if (files != null)
                {
                    for (file in files!!)
                    {
                        Log.i(TAG, "file =" + file.key + " : " + file.name)
                        DataLoaderUtil.setFileBuffer(file, wr, delimiter)
                    }
                }
                if (urlParam != null)
                {
                    val write = DataLoaderUtil.getParamsString(urlParam!!, delimiter)
                    wr.writeUTF(write)
                    Log.i(TAG, "dataUrl=$path?$write")
                }
                else
                {
                    Log.i(TAG, "dataUrl=" + path!!)
                }
                wr.writeBytes("\r\n--$boundary--")
                wr.flush()
                wr.close()


            } else {

                uc.setRequestProperty("Content-type", contentType)

                if (urlParam != null) {

                    val sb = StringBuffer()
                    val keySet = urlParam!!.keys.toTypedArray()
                    var key: String
                    for (i in keySet.indices)
                    {
                        key = keySet[i]
                        if (i == keySet.size - 1)
                        {
                            sb.append(key).append("=").append(urlParam!![key])
                        }
                        else
                        {
                            sb.append(key).append("=").append(urlParam!![key]).append("&")
                        }
                    }

                    val wr = DataOutputStream(uc.outputStream)
                    val write = sb.toString()
                    wr.writeBytes(write)
                    wr.flush()
                    wr.close()
                    Log.i(TAG, "dataUrl=$path?$write")
                }
                else
                {
                    Log.i(TAG, "dataUrl=" + path!!)

                }
            }
            val responseCode = uc.responseCode
            Log.i(TAG, "responseCode =" + uc.responseMessage)
            if (responseCode == HttpURLConnection.HTTP_OK) ism = uc.inputStream
        }
        catch (e: Exception)
        {
            Log.e(TAG, "Error in parser run", e)
            return null
        }

        if (ism != null)
        {

            val baos = ByteArrayOutputStream()
            val byteBuffer = ByteArray(1024)
            var byteData: ByteArray? = null
            try
            {
                do
                {
                    var nLength = ism!!.read(byteBuffer, 0, byteBuffer.size)
                    baos.write(byteBuffer, 0, nLength)
                }
                while (nLength != -1)

            }
            catch (e: IOException)
            {
                Log.e(TAG, "Error in InputStream read ", e)
                return null
            }

            byteData = baos.toByteArray()
            responseStr = String(byteData!!)
            //Log.i(TAG, "response=" + responseStr);

        }
        uc?.disconnect()
        return responseStr
    }


    override fun onPostExecute(result: String?)
    {

        if (result == null) delegate?.onLoadErr(path) else delegate?.onCompleted(path, result)
        super.onPostExecute(result)
    }

    companion object
    {
        private val TAG = "dataLoader"
    }

}//package

internal interface DataLoaderDelegate
{
    fun onCompleted(path: String?, result: String)
    fun onLoadErr(path: String?)
}
